numba/array_analysis.py

```diff
@@ -0,0 +1,718 @@
+            for (i,corr) in enumerate(rhs_corr):
+                # if corr unknown or new
+                if corr==-1 or corr not in self.class_sizes.keys():
+                    # generate size call nodes for this dimension
+                    nodes = self._gen_size_call(assign.target, i)
+                    size_calls += nodes
+                    assert isinstance(nodes[-1], ir.Assign)
+                    size_var = nodes[-1].target
+                    if corr!=-1:
+                        self.class_sizes[corr] = [size_var]
+                    self.array_size_vars[lhs][i] = size_var
+                else:
+                    # reuse a size variable from this correlation
+                    # TODO: consider CFG?
+                    self.array_size_vars[lhs][i] = self.class_sizes[corr][0]
+
+        #print(self.array_shape_classes)
+        return size_calls
+
+    def _gen_size_call(self, var, i):
+        out = []
+        ndims = self._get_ndims(var.name)
+        # attr call: A_sh_attr = getattr(A, shape)
+        shape_attr_call = ir.Expr.getattr(var, "shape", var.loc)
+        attr_var = ir.Var(var.scope, mk_unique_var(var.name+"_sh_attr"+str(i)), var.loc)
+        self.typemap[attr_var.name] = types.containers.UniTuple(types.int64, ndims)
+        attr_assign = ir.Assign(shape_attr_call, attr_var, var.loc)
+        out.append(attr_assign)
+        # const var for dim: $constA0 = Const(0)
+        const_node = ir.Const(i, var.loc)
+        const_var = ir.Var(var.scope, mk_unique_var("$const"+var.name+str(i)), var.loc)
+        self.typemap[const_var.name] = types.int64
+        const_assign = ir.Assign(const_node, const_var, var.loc)
+        out.append(const_assign)
+        # get size: Asize0 = A_sh_attr[0]
+        size_var = ir.Var(var.scope, mk_unique_var(var.name+"size"+str(i)), var.loc)
+        self.typemap[size_var.name] = types.int64
+        getitem_node = ir.Expr.static_getitem(attr_var, i, const_var, var.loc)
+        self.calltypes[getitem_node] = None
+        getitem_assign = ir.Assign(getitem_node, size_var, var.loc)
+        out.append(getitem_assign)
+        return out
+
+    # lhs is array so rhs has to return array
+    def _analyze_rhs_classes(self, node):
+        if isinstance(node, ir.Arg):
+            assert self._isarray(node.name)
+            return self._add_array_corr(node.name)
+        elif isinstance(node, ir.Var):
+            return copy.copy(self.array_shape_classes[node.name])
+        elif isinstance(node, (ir.Global,ir.FreeVar)):
+            # XXX: currently, global variables are frozen in Numba (can change)
+            if isinstance(node.value, np.ndarray):
+                shape = node.value.shape
+                out_eqs = []
+                for c in shape:
+                    new_class = self._get_next_class_with_size(c)
+                    out_eqs.append(new_class)
+                return out_eqs
+        elif isinstance(node, ir.Expr):
```

Comment:
Good catch! I added support for it. Since array analysis is called after `arrayexpr` pass, it was working already but it's good to be general.